src/js/utils/date-utils.js

```diff
@@ -0,0 +1,283 @@
+example: Jan 01, 2016 15:30 UTC
+python: "%b %d, %Y %H:%M"
+*/
+
+var DATE_FORMATS = {
+    DEFAULT_SHORT_DATE_FORMAT: 'll',
+    DEFAULT_LONG_DATE_FORMAT: 'LLLL',
+    DEFAULT_TIME_FORMAT: 'LTS z',
+    DEFAULT_DATE_TIME_FORMAT: 'll HH[:]mm z'
+};
+
+
+(function(define) {
+    'use strict';
+    define([
+        'jquery',
+        'edx-ui-toolkit/js/utils/string-utils',
+        'moment',
+        'momenttimezone'
+    ], function(
+        $,
+        StringUtils,
+        moment,
+        momentTZ
+        ) {
+        var DateUtils, isValid, dateObject, userTimezone, localizeTime, timeDisplay, localize;
+
+        /**
+         * Helper function to determine variable/context validity.
+         *
+         * @param {string} candidateVariable to test
+         * @returns {boolean}
+         */
+        isValid = function(candidateVariable) {
+            var valid = true;
+            if (candidateVariable === undefined) {
+                valid = false;
+            } else if (candidateVariable === 'undefined') {
+                valid = false;
+            } else if (candidateVariable === '') {
+                valid = false;
+            } else if (candidateVariable === 'Invalid date') {
+                valid = false;
+            } else {
+                valid = true;
+            }
+            return valid;
+        };
+
+        /**
+         * Attempt to convert a string to mutable 'moment' datetime object.
+         *
+         * The input datetimeString can be just a date, a date and time,
+         * or a complex date/time string with TZ offset. This is fairly wide-open, as
+         * many strings will satisfy this requirement.
+         * (e.g. 01-01-16 and Jan 1, 2016, and 2016-01-01 01:00:00+00:00 will work)
+         *
+         * The recommendation is to use a fully informed string, e.g. '2016-01-01 01:00:00+00:00',
+         * to limit possible date, time, and timezone ambiguities.
+         *
+         * @param {string} datetimeString. A string representation of a datetime.
+         * @returns {object} A moment.js UTC datetime object.
+         *
+        */
+        dateObject = function(datetimeString) {
+            var utcDateObject;
+            if (isValid(datetimeString)) {
+                utcDateObject = moment(datetimeString).utc();
+            } else {
+                utcDateObject = '';
+            }
+            return utcDateObject;
+        };
+
+        /**
+         * Attempt to determine user's timezone from preferences, then browser settings.
+         * Will default to UTC.
+         *
+         * @param {string} timezone 'Region/Population_Center' e.g. 'America/New_York'.
+         * If this is None or undefined, we attempt to determine the TZ from system/browser settings.
+         * @return {string} Will attempt to return, in this order, the entered, determined, and default (UTC)
+         * timezone.
+         *
+         */
+        userTimezone = function(timezone) {
```

Comment:
What is a more accurate name for this function?  Maybe `localTimezoneOrUTC`?
There is no "user", per say, that is involved here.  Since it includes "user" in its name I assumed it checked user preferences.  But this general purpose function simply guesses the local timezone and if not found, returns UTC.